pyi.py

```diff
@@ -303,6 +304,85 @@ def _is_object(node: ast.expr, name: str, *, from_: Container[str]) -> bool:
+
+
+def _is_type_or_Type(node: ast.expr) -> bool:
+    """
+    >>> import ast
+    >>> ast_node_for = lambda string: ast.parse(string).body[0].value
+    >>> _is_type_or_Type(ast_node_for('type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('Type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('builtins.type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('typing_extensions.Type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('typing.Type'))
+    True
+    """
+    if isinstance(node, ast.Name):
+        return node.id in {"type", "Type"}
+    if isinstance(node, ast.Attribute):
+        node_value = node.value
+        if not isinstance(node_value, ast.Name):
+            return False
+        node_value_id = node_value.id
+        attr = node.attr
+        return (node_value_id == "builtins" and attr == "type") or (
+            node_value_id in _TYPING_MODULES and attr == "Type"
+        )
+    return False
+
+
+def _is_PEP_604_union(node: ast.expr | None) -> TypeGuard[ast.BinOp]:
+    return isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitOr)
+
+
+def _is_None(node: ast.expr) -> bool:
+    # <=3.7: `BaseException | None` parses as BinOp(left=Name(id='BaseException'), op=BitOr(), right=NameConstant(value=None))`
+    # >=3.8: `BaseException | None` parses as BinOp(left=Name(id='BaseException'), op=BitOr(), right=Constant(value=None))`
+    # ast.NameConstant is deprecated in 3.8+, but doesn't raise a DeprecationWarning (and the isinstance() check still works)
+    return isinstance(node, ast.NameConstant) and node.value is None
+
+
+_sentinel = ast.expr()
+
+
+class ExitArgAnalysis(NamedTuple):
+    is_union_with_None: bool
+    non_None_part: ast.expr = _sentinel
+
+    def __repr__(self) -> str:
+        return (
+            f"ExitArgAnalysis("
+            f"is_union_with_None={self.is_union_with_None}, "
+            f"non_None_part={ast.dump(self.non_None_part)}"
+            f")"
+        )
+
+
+def _analyse_exit_method_arg(node: ast.BinOp) -> ExitArgAnalysis:
+    """Return a two-item tuple providing analysis of the annotation of an exit-method arg.
+
+    This function is only called if we already know
+    that `node` is a `BoolOp` and node.op is a `BitOr`.
+
+    >>> import ast
+    >>> ast_node_for = lambda string: ast.parse(string).body[0].value
+    >>> _analyse_exit_method_arg(ast_node_for('int | str'))
+    ExitArgAnalysis(is_union_with_None=False, non_None_part=expr())
+    >>> _analyse_exit_method_arg(ast_node_for('int | None'))
+    ExitArgAnalysis(is_union_with_None=True, non_None_part=Name(id='int', ctx=Load()))
+    >>> _analyse_exit_method_arg(ast_node_for('None | str'))
+    ExitArgAnalysis(is_union_with_None=True, non_None_part=Name(id='str', ctx=Load()))
+    """
+    if _is_None(node.left):
+        return ExitArgAnalysis(is_union_with_None=True, non_None_part=node.right)
+    if _is_None(node.right):
+        return ExitArgAnalysis(is_union_with_None=True, non_None_part=node.left)
```

Comment:
I'm pretty sure we already have code that turns the AST of `X | Y | Z` or `Union[X, Y, Z]` into a list `[X, Y, Z]`. Would it be useful here?

Reply:
We do, that logic is currently in `visit_BinOp`. But I don't think it's easily reusable here. In `visit_BinOp`, we need to be able to tackle variable-length BinOp unions. But here, we have to make sure that the union only has two elements -- so it makes sense to short-circuit early, I think.

Reply:
There's also the fact that the diff for this is already pretty complicated in parts, so I'd rather keep this PR isolated from the rest of the logic and then refactor other parts of the module afterwards. (E.g. the new `_is_type_or_Type` can be reused elsewhere, but I'd rather not do that in this PR.)